### PR TITLE
Various fixes + Pocket Loader & Placer Rewrite

### PR DESCRIPTION
**Pull request by CreepyCre**
### Current fixes:
- Rift targets now serialize correctly
- position printed out when teleporting via door now works correctly
- SchematicV1 redstone now works correctly

### Loader & Placer rewrite
The goal of the loader & placer rewrite is to allow for more pockets than just schematic based pockets.
It is already completely functional, current features are:
- loads pockets of different types from json
- can randomize pocket selection with dynamic weights
- use pockets conditionally based on depth
- Pockets can now do math. Example: depth based pocket size.

### The pocket system has been split into a PocketGenerator system and a VirtualPocket system.
1. PocketGenerator:
Responsible for generating a pocket.
2. VirtualPocket:
Represents a group of PocketGenerators. The VirtualPockets are pocket data structures that can select a PocketGenerator (randomly) based on PocketGenerationParameters supplied.

#### The current PocketGenerator types are:
1. SchematicGenerator:
Generates a pocket loaded from a schematic.
2. VoidGenerator:
Generates a void pocket. (no blocks, no walls)
3. ChunkGenerator (~70% done)
Generates a pocket based on the ChunkGenerator of a world specified. Size and offset may be configured.

#### The 3 general VirtualPockets types are:
1. VirtualPocketList
Is a weighted list of VirtualPockets
2. VirtualPocket selectors:
A type of VirtualPocket that reference other VirtualPockets and can choose their target based on PocketGenerationParameters supplied.
    1. DepthDependentSelector:
    Contains a List of String - VirtualPocket pairs, where the String is used as a regex to compare against depth. Will select first VirtualPocket whose regex matches the current depth. 
3. PocketGeneratorReference:
References PocketGenerators. May do so via id or tag (tags are planned atm)
    1. IdReference:
    References a PocketGenerator based on its id.

### Modifier system:
The new Modifiers can be applied to PocketGenerators or PocketGeneratorReferences and will be applied during pocket generation before the doors/ loottables are setup.
#### Current Modifiers:
1. ShellModifier:
Draws a shell based on the layers provided around a pocket.
2. DimensionalDoorModifier:
Places a dimensional door of given type at the given position relative to the pocket.

### CODE REVIEW

**Comment by CreepyCre** [comment]
reopened since there was an oppsie

File: src/main/java/org/dimdev/dimdoors/pockets/SchematicV2Handler.java [hunk 1]
```diff
@@ -46,63 +41,90 @@ public void load() {
...
+					loadCompound(directoryPath, directoryIdParts, loader);
+				}
+			} catch (IOException e) {
+				LOGGER.error("could not load pocket data in path " + path.toString() + " due to malformed json.", e);
+			}
+		} else if(Files.isRegularFile(path) && path.getFileName().toString().endsWith(".json")) {
+			String id = String.join(".", idParts);
+			try {
+				JsonObject json = GSON.fromJson(String.join("", Files.readAllLines(path)), JsonObject.class);
+				CompoundTag tag = CompoundTag.CODEC.decode(JsonOps.INSTANCE, json).getOrThrow(false, LOGGER::error).getFirst();
```
Comment (BoogieMonster1O1) [inline src/main/java/org/dimdev/dimdoors/pockets/SchematicV2Handler.java]: `JsonOps.INSTANCE.convertTo(NbtOps.INSTANCE, json)`

File: src/main/java/org/dimdev/dimdoors/pockets/generator/ChunkGenerator.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+		return tag;
+	}
+
+	@Override
+	public Pocket prepareAndPlacePocket(PocketGenerationParameters parameters) {
+		ServerWorld world = parameters.getWorld();
+		VirtualLocation sourceVirtualLocation = parameters.getSourceVirtualLocation();
+
+		int ChunkSizeX = ((this.size.getX() >> 4) + (this.size.getX() % 16 == 0 ? 0 : 1));
+		int ChunkSizeZ = ((this.size.getZ() >> 4) + (this.size.getZ() % 16 == 0 ? 0 : 1));
```
Comment (BoogieMonster1O1) [inline src/main/java/org/dimdev/dimdoors/pockets/generator/ChunkGenerator.java]: `chunkSizeX` and `chunkSizeZ`

File: src/main/java/org/dimdev/dimdoors/world/ModDimensions.java [hunk 3]
```diff
@@ -50,9 +50,11 @@ public static boolean isPocketDimension(World world) {
         return isPocketDimension(world.getRegistryKey());
     }
 
-    public static boolean isPocketDimension(RegistryKey<World> type) {
-        System.out.println("Type: " + type);
+    public static boolean isPersonalPocketDimension(World world) {
+        return isPocketDimension(world.getRegistryKey());
```
Comment (BoogieMonster1O1) [inline src/main/java/org/dimdev/dimdoors/world/ModDimensions.java]: this doesnt seem right

File: src/main/java/org/dimdev/dimdoors/pockets/SchematicV2Handler.java (hunk 1 above)
Comment (CreepyCre) [reply src/main/java/org/dimdev/dimdoors/pockets/SchematicV2Handler.java]: you're a godsend

**Pull request merged**